### PR TITLE
Virtual scroll rework

### PR DESCRIPTION
**Pull request by varnastadeus**
This PR will try to address the issues regarding virtual scroll and dropdown itself

### CODE REVIEW

**Comment by varnastadeus** [comment]
https://github.com/ng-select/ng-select/issues/899

**Comment by EricsG95** [comment]
This is a real issues that I encounter, and it will really great to see this to get merged it. Thank you for your hardwork.

**Comment by varnastadeus** [comment]
check after merge #657

File: src/ng-select/items-list.ts [hunk 1]
```diff
@@ -310,6 +310,15 @@ export class ItemsList {
         }
     }
 
+    private get _lastMarkedIndex() {
+        const selectedIndex = this._filteredItems.indexOf(this.lastSelectedItem);
+        if (selectedIndex === -1) {
+            this._markedIndex = selectedIndex;
```
Comment (anjmao) [inline src/ng-select/items-list.ts]: 1. Better convert it to method as it does some search and other operations.

```ts
private _getLastMarkedIndex(): number {}
```
2.  Getter should not modify state as it does here with `this._markedIndex = selectedIndex;`

File: src/ng-select/ng-dropdown-panel.component.ts [hunk 2]
```diff
@@ -1,35 +1,35 @@
...
-import { WindowService } from './window.service';
-import { VirtualScrollService } from './virtual-scroll.service';
-import { takeUntil } from 'rxjs/operators';
-import { Subject, fromEvent, merge } from 'rxjs';
+import { NgOption } from './ng-select.types';
+import { isDefined } from './value-utils';
 
 const TOP_CSS_CLASS = 'ng-select-top';
 const BOTTOM_CSS_CLASS = 'ng-select-bottom';
+const SCROLL_SCHEDULER = typeof requestAnimationFrame !== 'undefined' ? animationFrameScheduler : asapScheduler;
```
Comment (anjmao) [inline src/ng-select/ng-dropdown-panel.component.ts]: From what I see reaquestAnimationFrame is already supported even on IE10 https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame

In which case can it be undefined?

File: src/ng-select/ng-select.component.ts [hunk 3]
```diff
@@ -733,14 +733,22 @@ export class NgSelectComponent implements OnDestroy, OnChanges, AfterViewInit, C
...
         }
-        this.dropdownPanel.scrollIntoTag();
+        this.dropdownPanel.scrollToTag();
+    }
+
+    private _onSelectionChanged() {
+        if (this.isOpen && this.multiple && this.appendTo) {
+            // Make sure items are rendered.
+            this._cd.detectChanges();
+            this.dropdownPanel.adjustDropdownPosition();
```
Comment (anjmao) [inline src/ng-select/ng-select.component.ts]: As this is already dropdownPanel maybe it make sense to drop dropdown name from method.

```ts
this.dropdownPanel.adjustPosition()
```

File: src/ng-select/ng-dropdown-panel.component.ts (hunk 2 above)
Comment (varnastadeus) [reply src/ng-select/ng-dropdown-panel.component.ts]: while using server-side rendering (angular universal) requestAnimationFrame will be undefined

**Pull request merged**